src/Simulation/Common/PortablePDBReader.cs

```diff
@@ -0,0 +1,260 @@
+        /// Tuple of strings such that full URL consists of their concatenation.
+        /// First part of URL is URL root for all files in PDB and second part is relative path to given file.
+        /// </summary>
+        public static Tuple<string,string> TryGetFileUrl(string pdbLocation, string fileName )
+        {
+            if (fileName == null) return null;
+
+            string prefix = null;
+            string rest = null;
+
+            Tuple<string, string>[] fileCorrespondence = GetRemappingInfromation(pdbLocation);
+            if (fileCorrespondence != null)
+            {
+                foreach (Tuple<string, string> replacement in fileCorrespondence)
+                {
+                    if (fileName.StartsWith(replacement.Item1))
+                    {
+                        rest = fileName.Replace(replacement.Item1, "");
+                        prefix = replacement.Item2;
+                        if (prefix.Contains(@"/") && rest.Contains(@"\"))
+                        {
+                            rest = rest.Replace(@"\", @"/");
+                        }
+                        break;
+                    }
+                }
+            }
+            return new Tuple<string, string>(prefix,rest);
+        }
+    }
+
+    public static class PortablePDBEmbeddedFilesCache
+    {
+        public const string lineMarkPrefix = ">>>";
+        public const int lineNumberPaddingWidth = 6;
+
+        [ThreadStatic]
+        private static Dictionary<string, Dictionary<string, string>> embeddedFiles = null;
+
+        public static Dictionary<string, string> GetEmbeddedFiles(string pdbLocation)
+        {
+            if (embeddedFiles == null)
+            {
+                embeddedFiles = new Dictionary<string, Dictionary<string, string>>();
+            }
+
+            Dictionary<string, string> embeddedFilesFromPath = null;
+            if (embeddedFiles.TryGetValue(pdbLocation, out embeddedFilesFromPath))
+            {
+                return embeddedFilesFromPath;
+            }
+            else
+            {
+                try
+                {
+                    embeddedFilesFromPath = PortablePdbSymbolReader.GetEmbeddedFiles(pdbLocation);
+                }
+                finally
+                {
+                    embeddedFiles.Add(pdbLocation, embeddedFilesFromPath);
+                }
+                return embeddedFilesFromPath;
+            }
+        }
+
+        public static string GetEmbeddedFileRange( string pdbLocation, string fullName, int lineStart, int lineEnd, bool showLineNumbers = false, int markedLine = -1, string markPrefix = lineMarkPrefix)
+        {
+            Dictionary<string, string> sourceToFile = GetEmbeddedFiles(pdbLocation);
+            if (sourceToFile == null) return null;
+            string source = sourceToFile.GetValueOrDefault(fullName);
+            if (source == null) return null;
+            string[] lines = Regex.Split(source, "\r\n|\r|\n");
```

Comment:
>lines [](start = 21, length = 5)

Would Reader approach be more efficient than storing 3 copies of the source file in memory?